server/src/main/java/com/hedera/block/server/persistence/StreamPersistenceHandlerImpl.java

```diff
@@ -134,4 +153,28 @@ public void onEvent(ObjectEvent<SubscribeStreamResponseUnparsed> event, long l,
             notifier.notifyUnrecoverableError();
         }
     }
+
+    private void handleBlockItems(final SubscribeStreamResponseUnparsed response)
+            throws ParseException, InterruptedException {
+        final List<BlockItemUnparsed> blockItems = response.blockItems().blockItems();
+        final BlockItemUnparsed firstItem = blockItems.getFirst();
+        if (firstItem.hasBlockHeader()) {
+            final long blockNumber = Preconditions.requireWhole(
+                    BlockHeader.PROTOBUF.parse(firstItem.blockHeader()).number());
```

Comment:
This is a little bit risky, as it's easy to copy/paste without thinking about the performance impact.
I would _recommend_ you retain the result _every_ time you parse a protobuf object, and copy values _separately_, in case we need more values later.

Reply:
Here, I intend to read once the block number in order to not read it again. It is true it will be needed later, I believe this block number should be read inside the writer once and to be able to be queried later when/if needed. It makes sense that the writer should know the block number it is concerned with and also makes sense to be able to provide that information to external callers. If you have meant something else entirely, please let me know, maybe I am misunderstanding your ask here.

Reply:
I'm just suggesting not to combine the two calls:
```suggestion
            final BlockHeader header = BlockHeader.PROTOBUF.parse(firstItem.blockHeader());
            final long blockNumber = Preconditions.requireWhole(header.number());
```